internal/graph/cursors.go

```diff
@@ -310,3 +324,197 @@ func combineCursors(cursor *v1.Cursor, toAdd *v1.Cursor) (*v1.Cursor, error) {
 		Sections:   append(slices.Clone(cursor.Sections), toAdd.Sections...),
 	}, nil
 }
+
+// withParallelizedStreamingIterableInCursor executes the given handler for each item in the items list, skipping any
+// items marked as completed at the head of the cursor and injecting a cursor representing the current
+// item.
+//
+// For example, if items contains 3 items, and the cursor returned was within the handler for item
+// index #1, then item index #0 will be skipped on subsequent invocation.
+//
+// The next index is executed in parallel with the current index, with its results stored in a CollectingStream
+// until the next iteration.
+func withParallelizedStreamingIterableInCursor[T any, Q any](
+	ctx context.Context,
+	ci cursorInformation,
+	name string,
+	items []T,
+	parentStream dispatch.Stream[Q],
+	concurrencyLimit uint16,
+	handler func(ctx context.Context, ci cursorInformation, item T, stream dispatch.Stream[Q]) error,
+) error {
+	// Check the cursor for a starting index, before which any items will be skipped.
+	startingIndex, err := ci.integerSectionValue(name)
+	if err != nil {
+		return err
+	}
+
+	itemsToRun := items[startingIndex:]
+	if len(itemsToRun) == 0 {
+		return nil
+	}
+
+	// Queue up each iteration's worth of items to be run by the task runner.
+	tr := newPreloadedTaskRunner(ctx, concurrencyLimit, len(itemsToRun))
+	stream, err := newParallelLimitedIndexedStream[Q](ctx, ci, parentStream, len(itemsToRun))
+	if err != nil {
+		return err
+	}
+
+	// Schedule a task to be invoked for each item to be run.
+	for taskIndex, item := range itemsToRun {
+		taskIndex := taskIndex
+		item := item
+		tr.add(func(ctx context.Context) error {
+			if ci.limits.hasExhaustedLimit() {
+				return nil
+			}
+
+			// Create an updated cursor referencing the current item's index, so that any items returned know to resume from this point.
+			currentCursor, err := ci.withOutgoingSection(name, strconv.Itoa(taskIndex+startingIndex))
+			if err != nil {
+				return err
+			}
+
+			// If not the first iteration, we need to clear incoming sections to ensure the iteration starts at the top
+			// of the cursor.
+			if taskIndex > 0 {
+				currentCursor = currentCursor.clearIncoming()
+			}
+
+			// Invoke the handler with the current item's index in the outgoing cursor, indicating that
+			// subsequent invocations should jump right to this item.
+			ictx, istream, icursor := stream.forTaskIndex(ctx, taskIndex, currentCursor)
+
+			err = handler(ictx, icursor, item, istream)
+			if err != nil {
+				return err
+			}
+
+			return stream.completedTaskIndex(taskIndex)
+		})
+	}
+
+	// NOTE: since branches can be canceled if they have reached limits, the context Canceled error is ignored here.
+	if err := tr.startAndWait(); !errors.Is(err, context.Canceled) {
```

Comment:
but they'll also return canceled if the parent context is canceled right? and in that case we do want to return

Reply:
that'll be handled at the parent call site though